plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -68,46 +105,110 @@ class bp_connection_manager {
-      try {
-         for (auto& entry : peers) {
+   void set_configured_bp_peers(const std::vector<std::string>& peers) {
+      for (const auto& entry : peers) {
+         try {
             auto comma_pos = entry.find(',');
             EOS_ASSERT(comma_pos != std::string::npos, chain::plugin_config_exception,
-                       "auto-bp-peer must consists an account name and server address separated by a comma token");
+                       "p2p-auto-bp-peer must consist of an account name and server address separated by a comma");
             auto         addr = entry.substr(comma_pos + 1);
             account_name account(entry.substr(0, comma_pos));
+            const auto& [host, port, type] = net_utils::split_host_port_type(addr);
+            EOS_ASSERT( !host.empty() && !port.empty(), chain::plugin_config_exception,
+                        "Invalid p2p-auto-bp-peer ${p}, syntax host:port:[trx|blk]", ("p", addr));
 
+            fc_dlog(self()->get_logger(), "Setting p2p-auto-bp-peer ${a} -> ${d}", ("a", account)("d", addr));
             config.bp_peer_accounts[addr]     = account;
             config.bp_peer_addresses[account] = std::move(addr);
-            fc_dlog(self()->get_logger(), "Setting auto-bp-peer ${a} -> ${d}", ("a", account)("d", config.bp_peer_addresses[account]));
+         } catch (chain::name_type_exception&) {
+            EOS_ASSERT(false, chain::plugin_config_exception,
+                       "the account ${a} supplied by --p2p-auto-bp-peer option is invalid", ("a", entry));
          }
-      } catch (eosio::chain::name_type_exception&) {
-         EOS_ASSERT(false, chain::plugin_config_exception, "the account supplied by --auto-bp-peer option is invalid");
       }
    }
 
    // Only called at plugin startup
    template <typename T>
    void for_each_bp_peer_address(T&& fun) const {
-      for (auto& [_, addr] : config.bp_peer_addresses) { fun(addr); }
+      fc::lock_guard g(gossip_bps.mtx);
+      for (const auto& bp_peer : gossip_bps.index) {
+         fun(bp_peer.server_address);
+      }
+   }
+
+   // Only called at plugin startup
+   void set_bp_producer_peers(const std::vector<std::string>& peers) {
+      for (const auto& entry : peers) {
+         try {
+            config.my_bp_peer_accounts.emplace(chain::name(entry));
+         } catch (chain::name_type_exception&) {
+            EOS_ASSERT(false, chain::plugin_config_exception,
+                       "the producer ${p} supplied by --p2p-producer-peer option is invalid", ("p", entry));
+         }
+      }
+   }
+
+   // Called at startup and when peer key changes
+   // empty modified_keys means to update all
+   void update_bp_producer_peers(const chain::controller& cc, const flat_set<account_name>& modified_keys, const std::string& server_address)
+   {
+      if (config.my_bp_peer_accounts.empty())
+         return;
+      fc::lock_guard gm(mtx);
+      fc::lock_guard g(gossip_bps.mtx);
```

Comment:
When locking multiple mutexes, I believe we should use [scoped_lock](https://en.cppreference.com/w/cpp/thread/scoped_lock) to avoid deadlocks. Unfortunately I don't believe we have a `fc` version of it.

Reply:
Correct, there is not an fc version. Could add it, but didn't think it was worth it. The important thing is to always do it in the same order.